pkg/providers/securitygroup/securitygroup.go

```diff
@@ -25,37 +25,47 @@ import (
 	"github.com/mitchellh/hashstructure/v2"
 	"github.com/patrickmn/go-cache"
 	"github.com/samber/lo"
+	"k8s.io/apimachinery/pkg/types"
 	"knative.dev/pkg/logging"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 
 	"sigs.k8s.io/karpenter/pkg/utils/pretty"
 
 	"github.com/aws/karpenter-provider-aws/pkg/apis/v1beta1"
 )
 
 type Provider interface {
-	List(context.Context, *v1beta1.EC2NodeClass) ([]*ec2.SecurityGroup, error)
+	List(context.Context, *v1beta1.EC2NodeClass) ([]v1beta1.SecurityGroup, error)
+	UpdateSecurityGroups(context.Context, *v1beta1.EC2NodeClass) ([]*ec2.SecurityGroup, error)
 }
 
 type DefaultProvider struct {
-	sync.Mutex
-	ec2api ec2iface.EC2API
-	cache  *cache.Cache
-	cm     *pretty.ChangeMonitor
+	sync.RWMutex
+	ec2api     ec2iface.EC2API
+	kubeClient client.Client
+	cache      *cache.Cache
+	cm         *pretty.ChangeMonitor
 }
 
-func NewDefaultProvider(ec2api ec2iface.EC2API, cache *cache.Cache) *DefaultProvider {
+func NewDefaultProvider(kubeClient client.Client, ec2api ec2iface.EC2API, cache *cache.Cache) *DefaultProvider {
 	return &DefaultProvider{
-		ec2api: ec2api,
-		cm:     pretty.NewChangeMonitor(),
+		ec2api:     ec2api,
+		kubeClient: kubeClient,
+		cm:         pretty.NewChangeMonitor(),
 		// TODO: Remove cache cache when we utilize the security groups from the EC2NodeClass.status
 		cache: cache,
 	}
 }
 
-func (p *DefaultProvider) List(ctx context.Context, nodeClass *v1beta1.EC2NodeClass) ([]*ec2.SecurityGroup, error) {
-	p.Lock()
-	defer p.Unlock()
+func (p *DefaultProvider) List(ctx context.Context, nodeClass *v1beta1.EC2NodeClass) ([]v1beta1.SecurityGroup, error) {
+	if err := p.kubeClient.Get(ctx, types.NamespacedName{Name: nodeClass.Name}, nodeClass); err != nil {
+		return nil, err
+	}
+	return nodeClass.Status.SecurityGroups, nil
+}
 
+// Use security groups provided by the nodeclass status
+func (p *DefaultProvider) UpdateSecurityGroups(ctx context.Context, nodeClass *v1beta1.EC2NodeClass) ([]*ec2.SecurityGroup, error) {
 	// Get SecurityGroups
```

Comment:
Do we need to `p.mu.Lock()` like we do with instance types and subnets? What's the risk of multiple NodeClasses bypassing the cache all at once and generating a lot more API calls than we expect?